app/lib/task/backend.dart

```diff
@@ -1024,6 +1027,40 @@ class TaskBackend {
       }
     });
   }
+
+  /// Returns the latest version of the [package] which has a finished analysis.
+  ///
+  /// Returns `null` if no such version exists.
+  Future<String?> latestFinishedVersion(String package) async {
+    final cachedValue =
+        await cache.latestFinishedVersion(package).get(() async {
+      for (final rt in acceptedRuntimeVersions) {
+        final key = PackageState.createKey(_db, rt, package);
+        final state = await dbService.lookupOrNull<PackageState>(key);
+        // skip states where the entry was created, but no analysis has not finished yet
+        if (state == null || state.hasNeverFinished) {
+          continue;
+        }
+        final finishedVersions = state.versions?.entries
+            .where((e) => e.value.finished)
+            .map((e) => e.key)
+            .toList();
+        if (finishedVersions == null || finishedVersions.isEmpty) {
+          continue;
+        }
+        if (finishedVersions.length == 1) {
+          return finishedVersions.single;
+        }
+        final bestVersion = finishedVersions
+            .map((e) => Version.parse(e))
+            .reduce((a, b) =>
+                compareSemanticVersionsDesc(a, b, true, true) <= 0 ? a : b);
+        return bestVersion.toString();
```

Comment:
Should we split this into a utility function called `latestVersion(Iterable<Version> versions)` so that it can have a few unit tests -- this is a great place to be off by one :D